merlin-sdk/src/main/java/gov/nasa/jpl/aerie/merlin/protocol/types/ValueSchema.java

```diff
@@ -170,37 +140,28 @@ public <T> T match(final Visitor<T> visitor) {
    * @param map Any set of named {@link ValueSchema}s.
    * @return A new {@link ValueSchema} representing a heterogeneous set of named {@link ValueSchema}s.
    */
-  public static ValueSchema ofStruct(final Map<String, ValueSchema> map) {
+  static ValueSchema ofStruct(final Map<String, ValueSchema> map) {
     for (final var v : Objects.requireNonNull(map).values()) Objects.requireNonNull(v);
     final var value = Map.copyOf(map);
-    return new ValueSchema() {
-      @Override
-      public <T> T match(final Visitor<T> visitor) {
-        return visitor.onStruct(value);
-      }
-    };
+    return new StructSchema(value);
   }
 
   /**
    * Creates a {@link ValueSchema} representing an {@link Enum} type.
    *
    * @return A new {@link ValueSchema} representing an {@link Enum} type.
    */
-  public static ValueSchema ofVariant(final List<Variant> variants) {
+  static ValueSchema ofVariant(final List<Variant> variants) {
     Objects.requireNonNull(variants);
-    return new ValueSchema() {
-      public <T> T match(final Visitor<T> visitor) {
-        return visitor.onVariant(variants);
-      }
-    };
+    return new VariantSchema(variants);
   }
 
-  public static final ValueSchema REAL = ofReal();
-  public static final ValueSchema INT = ofInt();
-  public static final ValueSchema BOOLEAN = ofBoolean();
-  public static final ValueSchema STRING = ofString();
-  public static final ValueSchema DURATION = ofDuration();
-  public static final ValueSchema PATH = ofPath();
+  ValueSchema REAL = new RealSchema(new Object());
```

Comment:
Not too happy about the seemingly useless `new Object()` here. It exists to allow each `record` to delegate its hash code, otherwise `ValueSchema.INT.hashCode() == ValueSchema.REAL.hashCode()` when both schema `record`s are defined like:
```java
record SomeTypeSchema() implements ValueSchema {
  ...
}
```
Without any members.

Reply:
But considering that we get a free `equals()` and `hashCode()` out of this I think the tradeoff may be worthwhile.

Reply:
Do I understand the problem correctly, in that records with the same elements, but different class names, have the same hashCode?

Reply:
It's okay if different `hashCode` implementations sometimes return the same thing. What's important is that `hashCode(x) == hashCode(y)` whenever `x.equals(y)`, which we get free for records.

Especially in the contexts where we use ValueSchema -- that is, approximately never as a key -- we shouldn't be concerned with this collision. The `new Objec()` is a poor trade.

Reply:
So you're saying `ValueSchema.INT.hashCode() == ValueSchema.REAL.hashCode()` should be acceptable? I don't think `ValueSchema.INT.equals(ValueSchema.REAL)` should yield `true`, so I would expect a unique hash code for these 2 schemas as well, you know?

Reply:
Hash functions, by definition, take a big space of elements and squish it down to a smaller space of numbers, because numbers are much easier to compare. Hash functions will always have collisions; a system that assumes hashes are unique is definitionally incorrect.

The right angle to take on `hashCode` is performance-based -- how evenly can we spread objects amongst numbers, so that we can filter a set down to a much smaller list of candidates rather than iterating over the entire set?

Reply:
The arrow points the other way - if they're equal, they _must_ have the same hashcode. That's the only rule. The converse doesn't need to hold (i.e. if they're not equal, the hashcodes may or may not be equal, it doesn't matter*)

*it matters for performance, but not correctness. In this case, a) we don't use ValueSchemas as keys very often, and b) even if we did, we'd be looking at number of collisions linear in the number of variants of `ValueSchema` (i.e. on the order of 10), which I doubt would have a measurable effect on performance

Reply:
> I don't think `ValueSchema.INT.equals(ValueSchema.REAL)` should yield `true`,

They don't. The implication goes the other way -- if the hashes are *distinct*, the objects *cannot* be equal. Nothing says that equal objects must have equal hashes.

Reply:
Thanks @Twisol @mattdailis, was getting myself confused with uniqueness/equality. I'll take the `Object` out of these records!